oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwsEncoder.java

```diff
@@ -0,0 +1,341 @@
+		Assert.notNull(headers, "headers cannot be null");
+		Assert.notNull(claims, "claims cannot be null");
+
+		JWK jwk = this.jwkSelector.apply(headers);
+		if (jwk == null) {
+			throw new JwtEncodingException(
+					String.format(ENCODING_ERROR_MESSAGE_TEMPLATE, "Failed to select a JWK signing key"));
+		}
+		else if (!StringUtils.hasText(jwk.getKeyID())) {
+			throw new JwtEncodingException(String.format(ENCODING_ERROR_MESSAGE_TEMPLATE,
+					"The \"kid\" (key ID) from the selected JWK cannot be empty"));
+		}
+
+		JWSSigner jwsSigner = this.jwsSigners.computeIfAbsent(jwk, (key) -> {
+			try {
+				return JWS_SIGNER_FACTORY.createJWSSigner(key);
+			}
+			catch (JOSEException ex) {
+				throw new JwtEncodingException(String.format(ENCODING_ERROR_MESSAGE_TEMPLATE,
+						"Failed to create a JWS Signer -> " + ex.getMessage()), ex);
+			}
+		});
+
+		// @formatter:off
+		JoseHeader.Builder headersBuilder = JoseHeader.from(headers)
+				.type(JOSEObjectType.JWT.getType())
+				.keyId(jwk.getKeyID());
+		JwtClaimsSet.Builder claimsBuilder = JwtClaimsSet.from(claims)
+				.id(UUID.randomUUID().toString());
+		// @formatter:on
+
+		this.jwtCustomizer.accept(headersBuilder, claimsBuilder);
+
+		headers = headersBuilder.build();
+		claims = claimsBuilder.build();
+
+		JWSHeader jwsHeader = JWS_HEADER_CONVERTER.convert(headers);
+		JWTClaimsSet jwtClaimsSet = JWT_CLAIMS_SET_CONVERTER.convert(claims);
+
+		SignedJWT signedJwt = new SignedJWT(jwsHeader, jwtClaimsSet);
+		try {
+			signedJwt.sign(jwsSigner);
+		}
+		catch (JOSEException ex) {
+			throw new JwtEncodingException(
+					String.format(ENCODING_ERROR_MESSAGE_TEMPLATE, "Failed to sign the JWT -> " + ex.getMessage()), ex);
+		}
+		String jws = signedJwt.serialize();
+
+		return new Jwt(jws, claims.getIssuedAt(), claims.getExpiresAt(), headers.getHeaders(), claims.getClaims());
+	}
+
+	/**
+	 * The default {@code com.nimbusds.jose.jwk.JWK} selector that matches on the
+	 * {@link JoseHeader#getAlgorithm() JWS algorithm}.
+	 */
+	public static final class DefaultJwkSelector implements Function<JoseHeader, JWK> {
+
+		private final Supplier<JWKSet> jwkSetProvider;
+
+		/**
+		 * Constructs a {@code DefaultJwkSelector} using the provided parameters.
+		 * @param jwkSetProvider a {@code Supplier} of
+		 * {@code com.nimbusds.jose.jwk.JWKSet}
+		 */
+		public DefaultJwkSelector(Supplier<JWKSet> jwkSetProvider) {
+			Assert.notNull(jwkSetProvider, "jwkSetProvider cannot be null");
+			this.jwkSetProvider = jwkSetProvider;
+		}
+
+		@Override
+		public JWK apply(JoseHeader headers) {
+			Assert.notNull(headers, "headers cannot be null");
+
+			JWSAlgorithm jwsAlgorithm = JWSAlgorithm.parse(headers.getAlgorithm().getName());
+			JWSHeader jwsHeader = new JWSHeader(jwsAlgorithm);
+			JWKSelector jwkSelector = new JWKSelector(JWKMatcher.forJWSHeader(jwsHeader));
+			List<JWK> jwks = jwkSelector.select(this.jwkSetProvider.get());
+			if (jwks.size() > 1) {
```

Comment:
> I'm assuming that the JWK source here is the same JWK source that an application would use for their `/jwks` authorization server endpoint

No. The JWK Source for the Authorization Server would behave differently than the one with Resource Server.

When a JWT needs to be signed, the JWK Source must return one `JWK` for signing. If it returns multiple, then how does it choose which one to use for signing? There is no metadata associated to the JWK indicating the "active" key. **NOTE:** This was the original thinking behind the design of the `ManagedKey`, as it would hold this metadata attribute indicating the active flag. Now that we removed `ManagedKey`, the JWK Source must return the one `active` (current) key used for signing.

The resource server could contain 2 keys that are active but it will narrow the selection process using a unique JOSE header, e.g. `kid`

Reply:
@jgrandja I agree with you, there can only be one signing key at a time. Is it possible to not require the caller code to provide the kid or alg in the header ? This could be guess from the signing key source if provided and will avoid the caller code to have access to the jwksource, parse it and extract header value which is already done by the encoder.

I think in most use case, it is the key that drive the kid and alg set in the header. The caller code just want to encode a jwt and don't bother which alg or key will be used. It is often generated outside and rolling. If the algorithm change, the caller code need to change if it has been hard coded like the sample given above `encoder.encode(Jwt.withAlgorithm(JwsAlgorithm.RS256).subject("subject"))`

Reply:
@scrocquesel NOTE: The code in this PR is out-of-date and needs to be synced with the code currently in the package [org.springframework.security.oauth2.client.endpoint](https://github.com/spring-projects/spring-security/tree/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint). The implementation `NimbusJwsEncoder` is used for Jwt Client Authentication introduced in `5.5`.

You will notice that all classes are `package-private` as we were not ready to expose `JwtEncoder` in `5.5` but we needed to add the Jwt Client Authentication feature, hence the reason for `package-private`.

If you have a need for the `JwtEncoder` API, it would be helpful if you could detail your requirements and we could consider exposing it in `5.6`

Reply:
Thanks, I didn't notice it was old code. Should I express needs in this issue ?

Reply:
Yes. Please add a new comment at bottom of this PR detailing your requirements and the need for the `JwtEncoder` API. The more detail the better. Thanks.